flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/AdaptedRestartPipelinedRegionStrategyNGFailoverTest.java

```diff
@@ -294,6 +294,44 @@ public void testFailGlobalIfErrorOnRestartTasks() throws Exception {
 		assertNotEquals(globalModVersionBeforeFailure, globalModVersionAfterFailure);
 	}
 
+	@Test
+	public void testCountingRestarts() throws Exception {
+		final JobGraph jobGraph = createStreamingJobGraph();
+		final ExecutionGraph eg = createExecutionGraph(jobGraph);
+
+		final Iterator<ExecutionVertex> vertexIterator = eg.getAllExecutionVertices().iterator();
+		final ExecutionVertex ev11 = vertexIterator.next();
+		final ExecutionVertex ev12 = vertexIterator.next();
+		final ExecutionVertex ev21 = vertexIterator.next();
+		final ExecutionVertex ev22 = vertexIterator.next();
+
+		// trigger task failure of ev11, vertices { ev11, ev21 } should be affected
+		ev11.getCurrentExecutionAttempt().fail(new Exception("Test Exception"));
+		assertEquals(1, eg.getNumberOfRestarts());
+
+		// finish the failover
+		manualMainThreadExecutor.triggerAll();
+		manualMainThreadExecutor.triggerScheduledTasks();
+		ev21.getCurrentExecutionAttempt().completeCancelling();
+		manualMainThreadExecutor.triggerAll();
+		manualMainThreadExecutor.triggerScheduledTasks();
+
+		// trigger task failure of ev12, vertices { ev12, ev22 } should be affected
```

Comment:
Is it needed to introduce a second locally recoverable failure? Isn't it enough to fail `ev11`, if we want to test that `executionGraph.incrementRestarts();` is invoked?

Reply:
done.